test/functional/owners_controller_test.rb

```diff
@@ -0,0 +1,213 @@
+        should redirect_to("ownership index") { rubygem_owners_path(@rubygem) }
+        should "remove the ownership record" do
+          assert @rubygem.owners_including_unconfirmed.include?(@last_ownership.user)
+        end
+        should "should flash error" do
+          assert_equal "Owner cannot be removed!", flash[:alert]
+        end
+      end
+    end
+
+    context "on GET to resend confirmation" do
+      setup do
+        @new_owner = create(:user)
+        @ownership = create(:ownership, :unconfirmed, rubygem: @rubygem, user: @new_owner)
+        sign_in_as(@new_owner)
+      end
+      context "with correct params" do
+        setup do
+          get :resend_confirmation, params: { rubygem_id: @rubygem.name, handle: @new_owner.display_id }
+        end
+
+        should redirect_to("rubygem show") { rubygem_path(@rubygem) }
+        should "set success notice flash" do
+          success_flash = "A confirmation mail has been re-sent to #{@new_owner.handle}'s email"
+          assert_equal success_flash, flash[:notice]
+        end
+      end
+
+      context "with incorrect params" do
+        should "show 404 error page if gem not found" do
+          get :resend_confirmation, params: { rubygem_id: "no_gem", handle: @new_owner.display_id }
+          assert_response :not_found
+        end
+
+        should "resend to signed in user irrespective of handle" do
+          get :resend_confirmation, params: { rubygem_id: @rubygem.name, handle: "no_handle" }
+          assert_redirected_to @rubygem
+          success_flash = "A confirmation mail has been re-sent to #{@new_owner.handle}'s email"
+          assert_equal success_flash, flash[:notice]
+        end
+
+        should "show alert if save failed" do
+          Ownership.any_instance.stubs(:save).returns(false)
+          get :resend_confirmation, params: { rubygem_id: @rubygem.name, handle: @new_owner.display_id }
+          assert_redirected_to @rubygem
+          success_flash = "Something went wrong. Please try again."
+          assert_equal success_flash, flash[:alert]
+        end
+      end
+    end
+  end
+
+  context "When user not logged in" do
+    setup do
+      @user = create(:user)
+      @rubygem = create(:rubygem)
+    end
+
+    context "on GET to confirm" do
+      setup do
+        @ownership = create(:ownership, :unconfirmed, user: @user, rubygem: @rubygem)
+      end
+
+      context "when token has not expired" do
+        should "confirm ownership" do
+          get :confirm, params: { rubygem_id: @rubygem.name, token: @ownership.token }
+          @ownership.reload
+          assert @ownership.confirmed?
+          assert redirect_to("rubygem show") { rubygem_path(@rubygem) }
+          assert_equal flash[:notice], "You are added as an owner to #{@rubygem.name} gem!"
+        end
+      end
+
+      context "when token has expired" do
+        setup do
+          @ownership.token_expires_at = Time.current - 2.hours
+          @ownership.save
```

Comment:
This setup would be cleaner if we were not sharing setup (line 152) with the above context (line 157).

Reply:
Do you mean doing 

`@ownership = create(:ownership, :unconfirmed, user: @user, rubygem: @rubygem, token_expires_at: Time.current - 2.hours)`

in setup (L:152)?
token_expires_at isn't defined in factory..

Reply:
> token_expires_at isn't defined in factory..

Add it please

Reply:
`token_expires_at` is being set by before_create hook.
I think using `travel_to` makes sense here.